packages/graphql-codegen-cli/src/utils/patterns.ts

```diff
@@ -0,0 +1,308 @@
+   */
+  return ({ path: absolutePath }: { path: string }) => {
+    if (!isAbsolute(absolutePath)) {
+      throw new Error('shouldRebuild trigger should be called with absolute path');
+    }
+
+    const path = relative(process.cwd(), absolutePath);
+    const shouldRebuild = localMatchers.some(matcher => matcher(path));
+    return shouldRebuild;
+  };
+};
+
+/**
+ * Create the pattern set for the "global" (top level) config.
+ *
+ * In the `shouldRebuild` algorithm, any of these watch patterns will take
+ * precedence over local configs, and any schemas and documents patterns will be
+ * mixed into the pattern set of each local config.
+ */
+export const makeGlobalPatternSet = (initialContext: CodegenContext) => {
+  const config: Types.Config & { configFilePath?: string } = initialContext.getConfig();
+
+  return {
+    watch: sortPatterns([
+      ...(typeof config.watch === 'boolean' ? [] : normalizeInstanceOrArray<string>(config.watch ?? [])),
+      relative(process.cwd(), initialContext.filepath),
+    ]),
+    schemas: sortPatterns(makePatternsFromSchemas(normalizeInstanceOrArray<Types.Schema>(config.schema))),
+    documents: sortPatterns(
+      makePatternsFromDocuments(normalizeInstanceOrArray<Types.OperationDocument>(config.documents))
+    ),
+  };
+};
+
+/**
+ * Create the pattern set for a "local" (output target) config
+ *
+ * In the `shouldRebuild` algorithm, any of these watch patterns will take
+ * precedence over documents or schemas patterns, and the documents and schemas
+ * patterns will be mixed into the pattern set of their respective gobal pattern
+ * set equivalents.
+ */
+export const makeLocalPatternSet = (conf: Types.ConfiguredOutput) => {
+  return {
+    watch: sortPatterns(normalizeInstanceOrArray(conf.watchPattern)),
+    documents: sortPatterns(
+      makePatternsFromDocuments(normalizeInstanceOrArray<Types.OperationDocument>(conf.documents))
+    ),
+    schemas: sortPatterns(makePatternsFromSchemas(normalizeInstanceOrArray<Types.Schema>(conf.schema))),
+  };
+};
+
+/**
+ * Parse a list of micromatch patterns from a list of documents, which should
+ * already have been normalized from their raw config values.
+ */
+const makePatternsFromDocuments = (documents: Types.OperationDocument[]): string[] => {
+  const patterns: string[] = [];
+
+  if (documents) {
+    for (const doc of documents) {
+      if (typeof doc === 'string') {
+        patterns.push(doc);
+      } else {
+        patterns.push(...Object.keys(doc));
+      }
+    }
+  }
+
+  return patterns;
+};
+
+/**
+ * Parse a list of micromatch patterns from a list of schemas, which should
+ * already have been normalized from their raw config values.
+ */
+const makePatternsFromSchemas = (schemas: Types.Schema[]): string[] => {
+  const patterns: string[] = [];
+
+  for (const s of schemas) {
+    const schema = s as string;
+    if (isGlob(schema) || isValidPath(schema)) {
+      patterns.push(schema);
+    }
+  }
+
+  return patterns;
+};
```

Comment:
This logic is exactly the same as it was previously. It's just been moved to this file.